### ISSUE TITLE
A better story of container metrics for runtime integration

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#27097]
Currently, kubelet gets container stats/metrics through cadvisor and expose them through the /stats endpoint.
This forces container runtimes to integrate with cadvisor first, before they can properly integrate with kubelet.

Now that we are defining the new container runtime interface for kubelet (#22964), we should define container metrics and associated methods in the new interfaces so that runtimes can integrate directly with kubelet. This will greatly reduce the integration pain.

Before doing that, we should audit existing cadvisor usage in kubelet and break down the cadvisor interface for different purposes. 

/cc @timstclair @vishh @mtaufen

**Comment by feiskyer** [comment]
+1

**Comment by yifan-gu** [comment]
cc @kubernetes/rktnetes-maintainers

**Comment by timstclair** [comment]
One thing to consider is that this will make adding additional metrics significantly more difficult. We may want to distinguish between required and nice-to-have metrics. It also depends on our future monitoring story, which is still up in the air (e.g. will these metrics be surfaced in the summary API and collected by Heapster? I suppose even if we went with a standalone solution, it could implement the client interface to collect the metrics).

**Comment by yujuhong** [comment]
+1 for a set of core metrics.

As for nice-to-have metrics, we can either 1) allow runtime to return empty stats if it doesn't support it, or 2) go with the standalone solution, in which case it's up to runtime to integrate with heapster or other components.

@timstclair @dchen1107 , what's the status about custom metrics and monitoring in general?

**Comment by timstclair** [comment]
I don't think any progress has been made since we decided not to prioritize it for 1.3, and I haven't seen anything about it on 1.4. @dchen1107 do you have any more context?

**Comment by alexbrand** [comment]
As @timstclair mentioned in https://github.com/google/cadvisor/issues/1394#issuecomment-234677894, it'd be great to keep Windows in mind when defining this interface.

**Comment by ncdc** [comment]
FYI @DirectXMan12

**Comment by tmrts** [comment]
@timstclair @yujuhong the next action item for this is to analyze & refactor the current cAdvisor usage in kubelet, is that correct?

**Comment by yujuhong** [comment]
> @timstclair @yujuhong the next action item for this is to analyze & refactor the current cAdvisor usage in kubelet, is that correct?

Yes, that'd give us a better idea of what metrics we need to support the current implementation.
I may not have time to get to this soon, so help will be welcome

**Comment by timstclair** [comment]
We've had a lot of discussions around separating "core" metrics that are required for internal kubernetes functionality from monitoring metrics that are user facing, and would be used for monitoring, alerting, etc. With this split, the CRI would only need to provide the core metrics (other metrics would be collected through alternate channels).

I'd really like to make this split before defining an API for it, but I think this is still a ways off.

**Comment by DirectXMan12** [comment]
Do HPA metrics count as "required for internal kubernetes functionality"?  How do you define what "core" is, especially if we open up the HPA to allow referring to arbitrary metrics?  Do we say "by default, the only metrics available in your cluster to scale on are CPU, memory, etc, and for anything else you need to write your own solution to gather and push to Heapster"?

**Comment by timstclair** [comment]
@DirectXMan12 And this is why I think it's a ways off :) In all seriousness though, HPA metrics are definitely part of the discussion. There needs to be a way to ingest them, whether or not they're a part of the core metrics exported by the runtime is another question.

**Comment by ConnorDoyle** [comment]
Just brainstorming, but how committed are folks to making metrics part of the CRI? Based on the above it sounds like: "HPA requires collecting some metrics, those metrics can come from multiple sources, one likely metrics source is the container runtime." Would it make sense to define a metrics source interface and then let container runtimes optionally implement it?

**Comment by vishh** [comment]
To clarify:
* `system` metrics refers to metrics provided by the OS like cpu, memory, disk usage etc.,
* `service` metrics refers to metrics provided by applications running in containers.
IIUC, `HPA metrics` here is referring to `service` metrics. 
I'd argue that service metrics should not belong in CRI. A limited set of system metrics (cpu, memory & disk) is required for kubelet and core k8s functionalities like scheduling. 
A more broader set of system metrics and service metrics can be made available to HPA (via heapster) through means that do not involve the runtime and kubelet. 
Requiring all runtimes to support a wide-array of system metrics and service metrics is unreasonable and unnecessary.

**Comment by DirectXMan12** [comment]
>  A more broader set of system metrics and service metrics can be made available to HPA (via heapster) through means that do not involve the runtime and kubelet.  Requiring all runtimes to support a wide-array of system metrics and service metrics is unreasonable and unnecessary.

I think we should come up with a story for how that's going to happen.  At the end of the day, it does not necessarily *have* to involve the CRI, but it may (I do like the idea of an optional interface, as @ConnorDoyle suggested), so I think we should work it out.  I'm wary of a scenario where every kubernetes cluster/platform has a different method for allowing users to expose "service" metrics, which would make applications not as easily portable between clusters.

**Comment by yujuhong** [comment]
> Just brainstorming, but how committed are folks to making metrics part of the CRI? Based on the above it sounds like: "HPA requires collecting some metrics, those metrics can come from multiple sources, one likely metrics source is the container runtime." Would it make sense to define a metrics source interface and then let container runtimes optionally implement it?

This is one possible route, but it requires kubelet figuring out whether runtime supports metrics (i.e., optional features) and whether kubelet's metrics library (a.k.a. cadvisor) supports the type of the runtime (e.g., hypervisor-based). It may make more sense to just push this down to the runtime.

**Comment by DirectXMan12** [comment]
> It may make more sense to just push this down to the runtime.

I had interpreted the suggestion as something like defining a well-known interface that had two different actions:

a) fetching metrics
b) indicating where to look for "service" metrics in a standard format from pods

So Kubelet can a) ask the container runtime for metrics, and b) say "by the way, expect to see 'service' metrics in format X in location Y for these pods in these places" (where "in location Y" might be "on port Z" or something like that).  The container runtime can then choose to implement those two pieces however it wants.  Runtimes would be free to use cAdvisor for this purpose, but would not be obligated to do so; use of cAdvisor would be behind the the metrics interface, so that Kubelet would not know about cAdvisor at all.

Fetching core metrics would be a separate interface (that would be mandatory), but all the metrics would be presented together by Kubelet for collection by Heapster.

**Comment by ConnorDoyle** [comment]
Or even more simply, configure Kubelet to look to one or more endpoints as metrics sources. One of them might happen to be a process that also implements the CRI interface. With gRPC you can stack multiple services. If guaranteeing system metrics are present is a concern, maybe the system metrics source endpoint could be a separate config value.

**Comment by yujuhong** [comment]
> Fetching core metrics would be a separate interface (that would be mandatory

Agreed.

> but all the metrics would be presented together by Kubelet for collection by Heapster.

This is debatable. A separate monitoring pipeline could allow more flexibility without changing the kubelet stats api.

> Or even more simply, configure Kubelet to look to one or more endpoints as metrics sources. One of them might happen to be a process that also implements the CRI interface.

If kubelet is going to be the hub of all custom metrics, this interface/api should be completely separate from the CRI.


I think we've all agreed that core metrics should be part of CRI, and we should aim for finishing that first.
The non-core metrics, whether be optional in CRI, or as a standalone interface, is worth a separate discussion.

**Comment by andrzej-k** [comment]
This is my first message on Kubernetes community forum so: Hello all :wave:

I think that from the point of view of telemetry solutions having an interface to integrate with Kubelet (to provide core and/or non-core metrics) would be a great solution. 
In our [kubesnap](https://github.com/intelsdi-x/kubesnap) project we integrated [snap](https://github.com/intelsdi-x/snap) with Kubernetes to demonstrate new features, like: custom metrics on the fly, support for application level metrics, metrics processing and publishing to different sinks, etc. 
Metrics delivered by snap where consumed by Heapster (we added new data source to Heapster), but since it seems that Heapster can handle only one data source at a time ([link](https://github.com/kubernetes/heapster/blob/master/metrics/sources/factory.go#L44)), we had to configure Heapster to talk to snap only. 
If we could integrate with Kubelet directly, no changes on Heapster side would be needed. That's probably obvious truth but, I feel that loose coupling of monitoring with Kubelet could allow for new features to show up quicker, just my 2 cents...

**Comment by yujuhong** [comment]
I discussed with @dchen1107 offline. Below is the gist of our discussion.

 1. Core metrics is what CRI will be focus on, at least in the first version.
 2. There are concerns about extra overhead added by querying the runtime for container metrics, as opposed to using the built-in cadvisor that provides summary of all containers. 

Since cadvisor works for most (non-VM-based) runtimes, we can still get container level information from it. There are a few options to implement this.
 1. Define a new `MetricsService` interface. User can pass the endpoint at kubelet startup with `--metrics-endpoint=...`. If this flag is not set, by default, kubelet will use cadvisor (which also implements the `MetricsService`) for all container stats.
 2. Add container metrics calls in the current `RuntimeService` interface. Ask each runtime to register what features it supports at startup. If a runtime does not support metrics, kubelet will use cadvisor instead.

Option (1) favors composability and avoid dealing with optional runtime features for now. Option (2) is more flexible and can be used in a broader scenario.

My concern is that option (2), though might be needed in the long run for features such as apparmor, may not be suitable for declaring that the runtime does not support a set of methods in the interface. Another concern is that the potential abuse of feature registration, which may require more thoughts before implementing.

**Comment by yujuhong** [comment]
https://github.com/kubernetes/kubernetes/pull/29963#issuecomment-237269565

**Comment by vishh** [comment]
> Define a new MetricsService interface. User can pass the endpoint at kubelet startup with --metrics-endpoint=.... If this flag is not set, by default, kubelet will use cadvisor (which also implements the MetricsService) for all container stats.

This option will require cAdvisor understanding all runtimes and pods as a first class concept. The way we handle docker and rkt is not extensible to many other runtimes IMHO. If we go with this approach, we need to invest in cAdvisor to make it a general purpose metrics source for Kubelet.

I still think that having the runtime link in appropriate cAdvisor libraries to expose system metrics for containers directly via the CRI might be an optimal approach!

**Comment by DirectXMan12** [comment]
> Define a new MetricsService interface. User can pass the endpoint at kubelet startup with --metrics-endpoint=.... If this flag is not set, by default, kubelet will use cadvisor (which also implements the MetricsService) for all container stats.

I don't think you can safely have a default like that, since not all runtimes are supported by cAdvisor (and there could always be runtimes that cAdvisor doesn't know about).

I think separating the MetricsService from the choice of runtime is missing that metrics gathering is somewhat closely tied to the technologies behind the runtime.  Sure, there might be a single runtime that could support multiple gatherers, and there might be a single gatherer that supports multiple runtimes (e.g. cAdvisor), but the runtime and the gatherer are never truly independent.  I think it makes much more sense to just let the runtimes say "this is my metrics endpoint, go find metrics here".

**Comment by yujuhong** [comment]
> This option will require cAdvisor understanding all runtimes and pods as a first class concept. The way we handle docker and rkt is not extensible to many other runtimes IMHO. If we go with this approach, we need to invest in cAdvisor to make it a general purpose metrics source for Kubelet.

Do not disagree with that (cadvisor not being extensible). If any new runtime integration requires changes in the cadvisor codebase, it'd be a big no to me. Some questions
- What runtime-specific knowledge does cadvisor need for core container metrics (cpu, mem)? Assume pods do not need to be first class concept and kubelet can map container to pods itself.
- For certain things (imageFS path), could kubelet get this information and pass it to cadvisor?
- How much work is it to make cadvisor more general purpose?

> I still think that having the runtime link in appropriate cAdvisor libraries to expose system metrics for containers directly via the CRI might be an optimal approach!

I don't oppose that, and I think I've brought it up before. @dchen1107, you had concerns about performance and reliability about offloading the metrics work to container runtime completely.  WDYT?

**Comment by yujuhong** [comment]
> I don't think you can safely have a default like that, since not all runtimes are supported by cAdvisor (and there could always be runtimes that cAdvisor doesn't know about).

It doesn't have to be a "default". We can also make it an explicit decision that user has to specify.

> 
> I think separating the MetricsService from the choice of runtime is missing that metrics gathering is somewhat closely tied to the technologies behind the runtime. Sure, there might be a single runtime that could support multiple gatherers, and there might be a single gatherer that supports multiple runtimes (e.g. cAdvisor), but the runtime and the gatherer are never truly independent. I think it makes much more sense to just let the runtimes say "this is my metrics endpoint, go find metrics here".

I agree that metrics are more tightly coupled with the runtime, but as you said, it's also possible for gatherer to support multiple runtimes. It's possible that the runtime doesn't even know where the gatherer is. 
I don't see much difference of setting --metrics-endpoint explicitly and asking the runtime where the metrics endpoint is, but the former allows the scenarios where runtimes are unaware of the gatherer.

**Comment by timstclair** [comment]
> How much work is it to make cadvisor more general purpose?

The issue is that cAdvisor currently supports running standalone outside of Kubernetes. As long as we support that usecase, we shouldn't tightly couple Kubernetes concepts.

>  don't see much difference of setting --metrics-endpoint explicitly and asking the runtime where the metrics endpoint is, but the former allows the scenarios where runtimes are unaware of the gatherer.

I lost you... Why do we want the kubelet to gather metrics from an endpoint? Isn't that the purpose of the ContainerRuntimeInterface (to provide methods / endpoints)?

**Comment by yujuhong** [comment]
> I lost you... Why do we want the kubelet to gather metrics from an endpoint? Isn't that the purpose of the ContainerRuntimeInterface (to provide methods / endpoints)?

We provide the grpc implemention (i.e., RuntimeService and ImageManagerService) through endpoints. The comment was just saying that instead of letting runtime say "I don't support some of the grpc calls in RuntimeService, we can choose to define another endpoint.

> The issue is that cAdvisor currently supports running standalone outside of Kubernetes. As long as we support that usecase, we shouldn't tightly couple Kubernetes concepts.

I didn't mean Kubernetes concepts. All I need is a container ID and the cpu/mem stats associated with it.

**Comment by dchen1107** [comment]
Ideally, none of the core metrics should get from the container runtime at all. The core metrics I am referring here is the set of metrics required for admission control / QoS controlling including 1) the initial feasibility-checking done by the default scheduler and Kubelet 2) the eviction done by Kubelet when there is a resource pressure. In this case, the set of core metrics is include the follow metrics only:

* Node level metrics: CPU, Memory, rootfs, etc.
* Pod level metrics: CPU, memory, rootfs, volumes, etc. 
  * Since Kubernetes 1.3 release, Kubernetes switch to support pod QoS (https://github.com/kubernetes/kubernetes/blob/master/docs/design/resource-qos.md) instead of per-resource per-container QoS. 
  * Per pod cgroup is under development today, and hopefully we can have it done for 1.4 release. Otherwise 1.5 release for sure. 

All core metrics I listed are collected / should collected by Kubelet through:

* linked in cAdvisor library: example: per cgroup cpu, memory, disk usage, inode, etc. 
* other plugin system / library: example: volume usage from volumemgr etc. 

All other metrics are populated today through Kubelet / cAdvisor are not part of the core metrics, and they should retrieve through a separate monitoring pipeline. Standalone cAdvisor (https://github.com/kubernetes/kubernetes/issues/18770) is one of the approaches what we agreed upon to provide the monitoring metrics back then. But this can be replaced by other solution, for example snap (@ConnorDoyle). The monitoring metrics API of Kubernetes node defined at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go includes:
 
* per container stats
* network, disk io stats etc. Those stats are not required for today's QoS controlling system. But we can add them back to core metrics later if requires. 
* application-level custom metrics
* etc. 

If we agree upon or still agreed upon above overall re-architecture from the high level, we can have

* No core metrics are required from container runtime at all in the future. 
* Easily expand today's monitoring API to include more stats without destabilizing Kubelet, the node agent. 
* ...

cc/ @bgrant0607 @thockin @derekwaynecarr @philips for overall re-architecture
@davidopp @fgrzadkowski for master metrics API, heapster refactory, and HPA support
@smarterclayton @ncdc for HPA and custom metrics 

But the above is describing an ideal world, not what we can easily have in next two releases. To unblock CRI work and minimize the moving components in our eco-system concurrently, I think we still require Kubelet provides all monitoring metrics defined at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go This implies the first version of container runtime interface & implementation(s) have to provide the following metrics:

* NodeStats.Runtime.ImageFs: We already decided the current CRI implements image service, it is naturally fit to ask runtime serving image related metrics today. This is decided in today's interface.
* PodStats.NetworkStats: This is still under debate. If we decided to punt the network plugging work to the runtime, I think the runtime should provide those stats. I am not sure how useful the stats are, and who is using them. If no one is using, we can ignore them for now until we have the separate monitoring pipeline. 
* Podstats.ContainerStats[]: Today the pod level metrics is derived from the containers' stats. 
  * Soon pod level cgroup can provide both cpustats and memory stats for all runtimes (docker, rkt and hyper) for QoS controlling system.
  * The first version of docker and rkt support can use cAdvisor library to get cpustats and memorystats for QoS controlling system, and per-container monitoring.
  * The first version of hyper support, they can easily "fake" the cpustats and memorystats on their side since I don't think VM overcommit is required for now. Once hypervisor process is in cgroup, Kubelet provides real usage at pod level. Then it upto to hyper to decide if they want to export the container's usage for monitoring.
  * Rootfs and logs are also at container level, and is expected to be moved to pod level and resolved through project-id based disk quota module (ETA: TBD?). 

cc/ @kubernetes/goog-node

**Comment by vishh** [comment]
@dchen1107 The summary matches our offline discussions. Hence LGTM!

**Comment by yujuhong** [comment]
@dchen1107, thanks for the nice summary!

I think the conclusions drawn here are:
  1. Container metrics will eventually be plumbed through another pipeline that's independent of kubelet.
  2. Pod-level metrics will continue being provided by kubelet from cgroups, with the exception of network stats.

We need to extend CRI to support `ImageFs` stats, and we *may* be able to drop pod network stats for now.
For container metrics, we can either add a temporary metrics api (option 2), or wait until all necessary work is completed (option 1).

***Option 1 delay supporting container metrics for all runtimes***
  - docker/rkt gets all metrics through existing implementation, i.e., built-in cadvisor. Metrics support for other runtime integration are delayed.
  - complete the pod-level cgroup work for pod stats
  - complete the standalone cadvisor work for container metrics 

***Option 2 add optional metrics interface to support other runtimes***
  - docker/rkt gets all metrics through existing implementation, i.e., built-in cadvisor.
  - ADD *optional* container metrics to CRI to support other runtimes. Refactor kubelet to use this alternate source of metrics.
  - complete the pod-level cgroup work for pod stats
  - complete the standalone cadvisor work for container metrics 
  - DELETE the container metrics from CRI, or just leave it there but not using it.

Any thoughts?

**Comment by thockin** [comment]
@yujuhong #1 is simpler - do you think it is viable?  Can we do #1 now and do #2 if and only if we have to?

**Comment by davidopp** [comment]
Splitting cAdvisor into a part that handles core metrics and is compiled into Kubelet, and a part that runs as a separate pod and handles everything else (including custom metrics) and is interchangeable with other collection agents, seems reasonable to me.

I didn't fully understand why container vs. pod was the criteria you are using to split non-core metrics vs. core metrics, but it sounds like it makes the CRI simpler, and I don't know enough to object. :)

In terms of what the scheduler needs,
1. short-term it needs per-pod usage information for usage-based scheduling (#12455)
2. longer-term it needs some kind of value like Borg reservation (Section 5.5 [here](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43438.pdf) for folks who aren't familiar with it) on a per-pod basis
3. It is conceivable that even longer-term (and this is very speculative), we might want to dynamically schedule containers into pods, in which case the scheduler would probably want usage and/or reservation on a per-container basis. But this is so speculative I do not think we should consider it now.

I assume that (2) will actually come through the separate monitoring pipeline, not from the core cAdvisor? In that case it is not quite correct to say that the core metrics provides all the information needed for scheduling.

**Comment by yujuhong** [comment]
> @yujuhong #1 is simpler - do you think it is viable? Can we do #1 now and do #2 if and only if we have to?

I think it's viable. Let's wait it out and see if other tasks will make progress in the meantime. I'm a bit unsure that the standalone cadvisor work will be picked up anytime soon though.

**Comment by dchen1107** [comment]
@thockin re: https://github.com/kubernetes/kubernetes/issues/27097#issuecomment-237755481 
Yes, Option 1 is the approach we chose for now and meanwhile
1. Pod-level cgroup support is under developement (https://github.com/kubernetes/kubernetes/issues/27204). Hopefully we can get it done for 1.4, but for sure in 1.5. We need this w / o CRI work anyway. 
2. Standalone cAdvisor (#18770) work will be picked up by someone in sig-node when we do 1.5 planning once we agreed on the high level (Looks like we do now). Not sure who yet. cc/ @matchstick 

@davidopp re: https://github.com/kubernetes/kubernetes/issues/27097#issuecomment-237962464
Thanks for providing the requirements from the scheduler perspective. Here are my thoughts on scheduling requirements for resource metrics:
1. Splitting core metrics vs. non-core metrics should cover both scheduler needs for 1. and 2. listed by you above
2. New CRI actually makes dynamically scheduling containers into pods possible since the new API separates container lifecycle from pod lifecycle. This is one of the main reasons I initiated this project.
3. To really unblock CRI work without any functionality & usability regression, node has to export monitoring metrics which is defined at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go The current reference implementation is standalone cAdvisor. And we need someone working on heapster to make sure it can retrieve the metrics from the daemon other than kubelet.

**Comment by bgrant0607** [comment]
Users specify resource requests and limits per container. Thus, container-level metrics will be needed for vertical auto-scaling, resource estimation, and predictive horizontal auto-scaling. Presumably HPA currently use container-level metrics, as well.

**Comment by yujuhong** [comment]
> Users specify resource requests and limits per container. Thus, container-level metrics will be needed for vertical auto-scaling, resource estimation, and predictive horizontal auto-scaling. Presumably HPA currently use container-level metrics, as well.

@bgrant0607, the conclusion drawn from this discussion is that container metrics should be handled in a separate pipeline (standalone cadvisor + heapster), where kubelet may not be involved.

**Comment by bgrant0607** [comment]
@yujuhong I'm fine with the metrics being surfaced via another component. I was just pointing out that the data will be required.

**Comment by yujuhong** [comment]
We've decided to punt the monitoring issues to after v1.5, in the hope that the monitoring pipeline will be more well-defined by then.

**Comment by michmike** [comment]
@yujuhong can you please involve me in the planning for the monitoring aspect of the container runtime interface? i would like to represent the windows-aspect of this work.

thanks!

**Comment by yujuhong** [comment]
#45614 added methods in CRI to get container stats. We didn't have time to plumb this through kubelet, so CRI runtimes cannot use it yet. I will file a separate issue for finishing that.
